Add table joins via Data.Discrimination

cfhammill opened the pull request:
Hi Anthony, thanks for the amazing package. I'm in the process of moving some of my workflows from R to haskell, and this has been a really useful bridge for me.

I'm interested in implementing features from R's tidyverse, one thing I missed from dplyr is ergonomic table joins. For this I borrowed @ekmett's discrimination package. In this PR I've add inner, outer, left, and right joins. The main challenge was getting coming up with Grouping instances for records, I accomplished this with some orphan instances, the rest was mostly mechanical.

I'm still a pretty new haskell programmer so any suggestions would be appreciated. 

Cheers,

Chris

acowley commented on src/Frames/Joins.hs:
```diff
@@ -0,0 +1,191 @@
+{-# LANGUAGE ConstraintKinds, DataKinds, FlexibleContexts, FlexibleInstances,
```
We need a module-level comment. It'd be great to cater to those familiar with SQL with a brief SQL translation, and to general Haskell programmers by, ideally, putting an example into English words.

cfhammill replied:
Definitely. Will add.

acowley commented on src/Frames/Joins.hs:
```diff
@@ -0,0 +1,191 @@
+import Data.Functor.Contravariant.Divisible
+import qualified Data.Text as T
+
+dropCols :: 
```
There are several undocumented helpers defined here. If we're sure they're only of interest to this module, we should add an explicit export list to the module and not include them. If they might be useful outside of this module, they should be documented. To clarify which definitions we expect to be most useful, we could either move the helpers into a `Frames.Joins.Internal` module, or we could re-export the main API definitions from the `Frames` module itself.

cfhammill replied:
I think `dropCols` could probably be of general use, it's essentially anti-select. The rest probably not. I will add an explicit exports list. Potentially `dropCols` could be added to exploration along-side select. Not attached to the name `dropCols`.

cfhammill replied:
Added explicit export to fix this

acowley commented on src/Frames/Joins.hs:
```diff
@@ -0,0 +1,191 @@
+  grouping = contramap T.unpack grouping
+
+-- | Perform an inner join operation on two frames
+-- matching on 
```
I think this comment was accidentally truncated at some point.

cfhammill replied:
oops, will fix.

acowley commented on src/Frames/Joins.hs:
```diff
@@ -0,0 +1,191 @@
+      proj2 y = rcast y :: Record fs
+      mergeFun r l = justsFromRec $ mergeRec cols r l 
+      mergeLeftEmpty l = justsFromRec l <+> nothingsFromRec (dropCols cols (head bs))
+      mergeRightEmpty r = nothingsFromRec (head as) <+> justsFromRec (dropCols cols r)
```
Is this use of `head` known to be safe? We should avoid using it if possible.

cfhammill replied:
Hmm, probably not, I guess I hadn't considered an empty left or right frame. If I special case empty frames would that be sufficient? Then head would be safe.

cfhammill replied:
Head is no longer necessary with `TypeApplications`

acowley commented on src/Frames/Joins.hs:
```diff
@@ -0,0 +1,191 @@
+justsFromRec = rmap (Just . getIdentity)
+
+nothingsFromRec :: Record fs -> Rec Maybe fs
+nothingsFromRec = rmap (const Nothing)
```
Out of curiosity, was not using `rpure Nothing` a deliberate choice? I can imagine that if you have the argument handy, it can make the call site cleaner, but perhaps `TypeApplications` could help with that (i.e. `rpure @rs Nothing`).

cfhammill replied:
Nope, just ignorance. I hadn't encountered TypeApplications yet, it seems like they would simplify things. I got most of my ideas from mimicking Melt.hs

acowley replied:
I'm trying to navigate shifting `vinyl` and `Frames` to use `TypeApplications` rather than `Proxy` arguments as everyone moves to GHC 8 and newer. It means things get a little fragmented, but I like feeling as though things are moving in a good direction 😀

cfhammill replied:
Understood, I will try switching the API to `TypeApplications`. It does seem like a good direction.

acowley commented on src/Frames/Joins.hs:
```diff
@@ -0,0 +1,191 @@
+    , RecVec rs2'
+    , RecVec (rs ++ rs2')
+    ) =>
+    proxy fs -- ^ A quasiquoter with shared columns to join on,
```
Any thoughts about using `TypeApplications` instead of the `Proxy` values?

cfhammill replied:
Fixed in new commits

acowley commented:
I added some superficial comments, but haven't yet had time to think about how it actually works! 😄

codygman commented:
Whoa, I look forward to finding how I implemented these compared to what I
made work in:

https://github.com/codygman/frames-diff/tree/4904ce1e16a7d62145ff9bcc55e9df008fdf4af8

On Feb 23, 2018 10:09 AM, "Chris Hammill" <notifications@github.com> wrote:

> *@cfhammill* commented on this pull request.
> ------------------------------
>
> In src/Frames/Joins.hs
> <https://github.com/acowley/Frames/pull/105#discussion_r170292544>:
>
> > +
> +innerJoin cols a b =
> +    toFrame $
> +    concat
> +    (inner grouping mergeFun proj1 proj2 (toList a) (toList b))
> +    where
> +      mergeFun = mergeRec cols
> +      proj1 x = rcast x :: Record fs
> +      proj2 y = rcast y :: Record fs
> +
> +
> +justsFromRec :: Record fs -> Rec Maybe fs
> +justsFromRec = rmap (Just . getIdentity)
> +
> +nothingsFromRec :: Record fs -> Rec Maybe fs
> +nothingsFromRec = rmap (const Nothing)
>
> Nope, just ignorance. I hadn't encountered TypeApplications yet, it seems
> like they would simplify things. I got most of my ideas from mimicking
> Melt.hs
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/acowley/Frames/pull/105#discussion_r170292544>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AANyN-IP6kthLuSWxn5O-4-YyyxwrcLmks5tXuKagaJpZM4SQRZi>
> .
>

cfhammill commented:
Ahh, I didn't see you had already implemented inner joins @codygman. Yours probably has advantages to mine in that it works with the pipe representation, mine needs to materialize everything. In my favour, If my understanding of rlens is right, I don't think your version can handle multiple join columns (please correct me if I'm wrong).

From a more theoretical perspective Data.Discrimination grouping is O(n + m) where I think the standard approach is O(nm). My benchmarking against `dplyr` has shown my implementation under-performing its promises. a 36k x 36k inner join takes about .600 ms, `dplyr` takes 10. The case is much worse for a 36k x  67 join, with mine taking 500ms and `dplyr` taking microseconds.

cfhammill commented:
Switched the whole API to `TypeApplications`, much cleaner now.

acowley commented:
The documentation for each function should include small example applications to make it clear that `TypeApplications` are required. The `Proxy` arguments had the nice property of providing an opportunity to hang some haddocks on them, which is good because they're really important arguments for these functions! 😜

cfhammill commented:
I added some more verbose documentation to the joins and fixed a problem with right and outer joins, the join keys were getting dropped when the left side was empty.

acowley commented:
It looks like we're closing in on this being ready to merge! Do you @cfhammill or @codygman or anyone else have any more thoughts on performance? Maybe a good thing to do would be to include a benchmark both of the Haskell join code and a simple Python program to compare to. That way we can own up to where we're slow, and give every contributor a running start to help figure out how to make it go faster.

cfhammill commented:
Ouch, fixing the right merge really trashed the performance of left and outer joins. Before I had been dropping key columns from the right table before merging, this doesn't make sense if the key is missing from the left table. To fix this I included an `rcast` at the end to the right column ordering, but this appears to be quite expensive. I think an alternate implementation would be to fill the left key columns with the right key, then drop the right key columns. Then there would be no column reordering. I'll give this a shot to see if I can get the benchmarks back down. The pandas version seems a fair bit slower than dplyr, all 6 test joins run in about a second, so about 8x faster than we were before the perf drop.

cfhammill commented:
@acowley is there a way to use lenses with `TypeApplications`?

acowley commented:
Do you mean the `vinyl` lenses? Not directly. We could add a module to `vinyl` that exports a Proxy-free API.

cfhammill commented:
Yeah, in `proxy` free mode it seems not simple to do getting and setting. Might be worth adding a module to `vinyl` if that's the way things are going. 

For now, I'm lightly stuck on how best to implement my suggestion above. Don't have much experience yet with modifying records. Not convinced it will help the performance issue but it's worth a try.

cfhammill commented:
Possibly false alarm on the performance drop, `TypeApplications` had no drop in performance on my laptop at least. The pandas demo takes 0.5 seconds on my laptop, the frames version takes 8 seconds.

acowley commented:
I don't have time to help with the optimization effort at the moment, but I don't want that to block things since slow and working is better than non-existant. Do you want to do any rebasing to help make this effort trackable for people reading the commit history?

cfhammill commented:
Ok great, I think for this a squash and merge would be fine, nothing in the history is particularly useful.

acowley commented:
Woohoo! 🙌

On to performance tuning!

cfhammill commented:
Hooray!

codygman commented:
Great job @cfhammil @acowley!

On Wed, Feb 28, 2018 at 9:10 PM, Chris Hammill <notifications@github.com>
wrote:

> Hooray!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/acowley/Frames/pull/105#issuecomment-369460286>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AANyNxCH7XVsAut-_P5DYRPOZwpktZtXks5tZ2aSgaJpZM4SQRZi>
> .
>

cfhammill commented:
Good news, I sped up the joins dramatically by switching the `Text` instance of grouping to use `hash` from `hashable` instead of `unpack`. Brought all 6 benchmarks down from 8s to 1.5s. 0.5s of which is IO. So optimistically 2-3x slower than pandas, not sure their IO/join ratio, with (fingers crossed) much better asymptotics.

Using hashing is a trick Kmett mentions in the Discrimination source. It could be worth it to insist that join columns be `Hashable` and use `hash` as the default implementation. Probably slower for numeric types but faster for anything more complicated. I'll open a new PR to discuss sometime this week.

ekmett commented:
Note: there is an asymptotic difference when using `Hashable` in that it will now have to touch every character of all of the strings to sort or discriminate rather than their longest common prefixes. This degrades the `O(LCP(S) + n log s)` performance bound to `O(|S| + n log s)`.

cfhammill commented:
Thanks @ekmett. Do you have a sense for when we will get a speed up from using `Hashable`? My impression now is that there's a grouping cost per character of the LCP, so we're looking for places where `n * LCP(S) * grouping cost < sum_n(|S_n|)`, is that right?

cfhammill commented:
Whoops I missed a `+ n * grouping cost ` in the rhs of the inequality above.

ekmett commented:
Well, for infinitely long random strings you have nice results about the expected length of the common prefixes to be logarithmic in the number of strings. Using that model as a conservative guide, you'd suffer asymptotically when the average length of strings is > than the log of the number of them you have.

But your supply of strings likely aren't completely random so you'll find this situation is even worse.

That said, these aren't particularly realistic conditions, but they do give a sense of the difference.

In general, discrimination looks at the least amount of structure necessary to break things up into (optionally ordered) equivalence classes stably and productively.

When you use hashing you lose this property entirely as you have to look at all of each of your structures to calculate the hash in the first place.

cfhammill commented:
Thanks, that makes sense. So I guess I might get better mileage out of accelerating the `Grouping` instance for `Text`. I'll give that a shot and see if I can get close to the `hash` performance for my small tests.

cfhammill commented:
Awesome, I've just done a bit of reading and I think I roughly get it. Let me know if there's anything I can do to help.
